unstable/logging/logging.go

```diff
@@ -106,13 +179,177 @@ func (l *Host[L]) f(severity diag.Severity, msg string) {
-func (l *Host[L]) Warn(msg string)  { l.f(diag.Warning, msg) }
-func (l *Host[L]) Error(msg string) { l.f(diag.Error, msg) }
+func (l *host[L]) StatusUntyped() any {
+	return l.Status()
+}
+
+func (l *host[L]) Debug(msg string) { l.f(diag.Debug, msg) }
+func (l *host[L]) Info(msg string)  { l.f(diag.Info, msg) }
+func (l *host[L]) Warn(msg string)  { l.f(diag.Warning, msg) }
+func (l *host[L]) Error(msg string) { l.f(diag.Error, msg) }
+
+// Directs any logs written using the tflog API in the given Context to the given output.
+func setupRootLoggers(ctx context.Context, output io.Writer) context.Context {
+	desiredLevel := parseTfLogEnvVar()
+	sdkLoggerOptions := makeLoggerOptions("sdk", desiredLevel, output)
+	ctx = context.WithValue(ctx, sdkKey, hclog.New(sdkLoggerOptions))
+	ctx = context.WithValue(ctx, sdkOptionsKey, sdkLoggerOptions)
+	providerLoggerOptions := makeLoggerOptions("provider", desiredLevel, output)
+	ctx = context.WithValue(ctx, providerKey, hclog.New(providerLoggerOptions))
+	ctx = context.WithValue(ctx, providerOptionsKey, providerLoggerOptions)
+	return ctx
+}
+
+// Choose the default level carefully: logs at this level or higher (more severe) will be shown to the user of Pulumi
+// CLI directly by default. Experimentally it seems that Info is too verbose, for example Cloudflare provider emits
+// routine authentication messages at INFO level.
+func defaultTFLogLevel() hclog.Level {
+	return hclog.Warn
+}
+
+func makeLoggerOptions(name string, level hclog.Level, output io.Writer) *hclog.LoggerOptions {
+	if level == hclog.NoLevel {
+		level = defaultTFLogLevel()
+	}
+	return &hclog.LoggerOptions{
+		Name:              name,
+		Output:            output,
+		Level:             level,
+		IndependentLevels: true,
+		IncludeLocation:   true,
+		TimeFormat:        " ",
+
+		// Empirically the value of 1 seems to work in the current Pulumi setup, @caller field now points to the
+		// file where the logging originates.
+		AdditionalLocationOffset: 1,
+	}
+}
+
+// Re-interprets structured JSON logs as calls against LogSink. To be used with SetupRootLoggers.
+func newLogSinkWriter(ctx context.Context, sink Sink) io.Writer {
+	desiredLevel := parseTfLogEnvVar()
+	return &logSinkWriter{
+		desiredLevel: desiredLevel,
+		ctx:          ctx,
+		sink:         sink,
+	}
+}
+
+type logSinkWriter struct {
+	desiredLevel hclog.Level
+	ctx          context.Context
+	sink         Sink
+}
+
+var _ io.Writer = &logSinkWriter{}
+
+func (w *logSinkWriter) Write(p []byte) (n int, err error) {
+	n = len(p)
+	if w.sink == nil {
+		return
+	}
+
+	raw := string(p)
+	level, raw := parseLevelFromRawString(raw)
+
+	if level < w.desiredLevel {
+		return
+	}
+
+	urn, raw := parseUrnFromRawString(raw)
+	severity := logLevelToSeverity(level)
+
+	err = w.sink.Log(w.ctx, severity, urn, raw)
+	return
+}
+
+func parseTfLogEnvVar() hclog.Level {
+	env, present := os.LookupEnv(tfLogEnvVar)
+	if !present {
+		return hclog.NoLevel
+	}
+	switch strings.ToUpper(env) {
```

Comment:
We don't actually care if `present` is true, so I think we can simplify this block.

```suggestion
	switch strings.ToUpper(os.Getenv(tfLogEnvVar)) {
```

Reply:
Slight preference for leaving it explicit. One day perhaps deploy static analysis that prohibits os.Getenv.